pkg/envoy/registry/services.go

```diff
@@ -27,34 +34,168 @@ func (e ExplicitProxyServiceMapper) ListProxyServices(p *envoy.Proxy) ([]service
 
 // KubeProxyServiceMapper maps an Envoy instance to services in a Kubernetes cluster.
 type KubeProxyServiceMapper struct {
-	KubeController k8s.Controller
+	kubeController k8s.Controller
+	kubeEvents     chan interface{}
+	servicesForCN  sync.Map // map[certificate.CommonName][]service.MeshService
 }
 
-// ListProxyServices maps an Envoy instance to a number of Kubernetes services.
-func (k *KubeProxyServiceMapper) ListProxyServices(p *envoy.Proxy) ([]service.MeshService, error) {
-	cn := p.GetCertificateCommonName()
+// NewKubeProxyServiceMapper initializes a KubeProxyServiceMapper with an empty cache.
+func NewKubeProxyServiceMapper(controller k8s.Controller) *KubeProxyServiceMapper {
+	return &KubeProxyServiceMapper{
+		kubeController: controller,
+		kubeEvents: events.GetPubSubInstance().Subscribe(
+			announcements.PodAdded,
+			announcements.PodUpdated,
+			announcements.PodDeleted,
+			announcements.ServiceAdded,
+			announcements.ServiceUpdated,
+			announcements.ServiceDeleted,
+		),
+	}
+}
 
-	pod, err := envoy.GetPodFromCertificate(cn, k.KubeController)
-	if err != nil {
-		return nil, err
+// Run starts updating the proxy-to-services cache based on k8s notifications.
+func (k *KubeProxyServiceMapper) Run(stop <-chan struct{}) {
+	// Populate the cache using the cluster's existing state. Otherwise if
+	// existing resources are not modified, no events will come through to
+	// update the cache.
+	for _, pod := range k.kubeController.ListPods() {
+		k.handlePodUpdate(pod)
```

Comment:
@eduser25 Is this what you mean by "abusing" these APIs since this does a lock/unlock for each pod? If so, would you say it's better to do a single lock before the loop and a single unlock after?

I'm pretty sure the answer is "yes" in both cases, just want to confirm my understanding.